libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c

```diff
@@ -0,0 +1,3476 @@
+						  CK_OBJECT_HANDLE_PTR pxObject )
+{
+#define MAX_PUBLIC_KEY_SIZE	3000
+	mbedtls_pk_context xMbedContext;
+	int lDerKeyLength;
+	CK_BYTE_PTR pxDerKey = NULL;
+	CK_KEY_TYPE xKeyType;
+	CK_RV xResult = CKR_OK;
+	CK_ATTRIBUTE_PTR pxLabel = NULL;
+	CK_ATTRIBUTE_PTR pxClass = NULL;
+	CK_OBJECT_HANDLE xPalHandle = CK_INVALID_HANDLE;
+	mbedtls_pk_init( &xMbedContext );
+	mbedtls_ecp_keypair * pxKeyPair;
+
+	prvGetKeyType( &xKeyType, pxTemplate, ulCount );
+	if( xKeyType == CKK_RSA )
+	{
+		xResult = CKR_ATTRIBUTE_TYPE_INVALID;
+	}
+
+	#if ( pkcs11configSUPPRESS_ECDSA_MECHANISM != 1 )
+		else if( xKeyType == CKK_EC ) /* CKK_EC = CKK_ECDSA. */
+		{
+			prvGetLabel( &pxLabel, pxTemplate, ulCount );
+
+			/* An mbedTLS key is comprised of 2 pieces of data- an "info" and a "context".
+				* Since a valid key was not found by prvGetExistingKeyComponent, we are going to initialize
+				* the structure so that the mbedTLS structures will look the same as they would if a key
+				* had been found, minus the private key component. */
+
+			/* If a key had been found by prvGetExistingKeyComponent, the keypair context
+				* would have been malloc'ed. */
+			pxKeyPair = pvPortMalloc( sizeof( mbedtls_ecp_keypair ) );
+
+			if( pxKeyPair != NULL )
+			{
+				/* Initialize the info. */
+				xMbedContext.pk_info = &mbedtls_eckey_info;
+
+				/* Initialize the context. */
+				xMbedContext.pk_ctx = pxKeyPair;
+				mbedtls_ecp_keypair_init( pxKeyPair );
+				mbedtls_ecp_group_init( &pxKeyPair->grp );
+				/*/ * At this time, only P-256 curves are supported. * / */
+				mbedtls_ecp_group_load( &pxKeyPair->grp, MBEDTLS_ECP_DP_SECP256R1 );
+			}
+			else
+			{
+				xResult = CKR_HOST_MEMORY;
+			}
+
+			xResult = prvCreateECPublicKey( &xMbedContext, &pxLabel, &pxClass, pxTemplate, ulCount, pxObject );
+		}
+	#endif /* if ( pkcs11configSUPPRESS_ECDSA_MECHANISM != 1 ) */
+	else
+	{
+		PKCS11_PRINT( ( "Invalid key type %d \r\n", xKeyType ) );
+		xResult = CKR_MECHANISM_INVALID;
+	}
+
+	if( xResult == CKR_OK )
+	{
+		/* Store the key.*/
+		pxDerKey = pvPortMalloc( MAX_PUBLIC_KEY_SIZE );
+
+		if( pxDerKey == NULL )
+		{
+			xResult = CKR_HOST_MEMORY;
+		}
+	}
+
+	if( xResult == CKR_OK )
+	{
+		lDerKeyLength = mbedtls_pk_write_pubkey_der( &xMbedContext, pxDerKey, MAX_PUBLIC_KEY_SIZE );
+
+		/* Do not clean up the mbedTLS key context here as it will be used in psa_save_object. */
+	}
+
+	if( xResult == CKR_OK )
+	{
+		xPalHandle = PKCS11_PSA_SaveObject( pxClass,
+											pxLabel,
+											pxDerKey + ( MAX_LENGTH_KEY - lDerKeyLength ),
+											lDerKeyLength,
+											&xMbedContext );
+
+		if( xPalHandle == CK_INVALID_HANDLE )
+		{
+			xResult = CKR_DEVICE_MEMORY;
+		}
+
+		mbedtls_pk_free( &xMbedContext );
+	}
+
+	if( xResult == CKR_OK )
+	{
+		xResult = prvAddObjectToList( xPalHandle, pxObject, pxLabel->pValue, pxLabel->ulValueLen );
+	}
+
+	if( pxDerKey != NULL )
+	{
+		vPortFree( pxDerKey );
+	}
+
+	return xResult;
+}
+
+CK_RV prvMbedTLS_Initialize( void )
```

Comment:
Does this code run on the main MCU side, or on the secure MCU? 

Would I be correct in assuming that the secure side is using mbed crypto?

Reply:
TF-M runs on a single ARMv8 core. Yes, it is correct that the secure side is using mbed crypto which is called by the PSA crypto service and is initialized by TF-M. The code here is part of the PKCS11 library which runs on the Non-Secure side. The Non-Secure side calls the mbedtls functions for parsing a key or certificate.

Reply:
This shim layer makes calls to mbedtls_xxx crypto functions (which are in mbedCrypto) to perform crypto operations. According to your comment above, the mbedCrypto code is in secure area and exposed via PSA Crypto Service. 
So how will the PKCS11 layer which is in non-secure area call the mbedtls_xxx crypto functions which are in secure area? Do you intend to expose all those functions also via the PSA Crypto service?

Reply:
Hi Michael,
The PKCS#11 shim layer running at the non-secure side calls the PSA Storage and Crypto APIs to the secure side which runs TF-M.
The implementation of the PSA Crypto API in TF-M is currently mbedcrypto. But it doesn't have to be mbedcrypto as the secure side backend or even TF-M itself. This shim layer can support any PSA API compliant implementation which may use SW or HW crypto engine.
Hope this helps to clarify the relationship between PSA Crypto API (service) and mbedcrypto.

Reply:
Hi David, 
Maybe I dont undertand the file partitioning, but if you look at "prvCreateEcPrivateKey()" there is a function in there called "mbedtls_mpi_read_binary()". Where does mbedtls_mpi_read_binary() reside? I know its in bignum.c, and when I used mbedCrypto it was part of that codebase. So the assumption I made is that that function is in the secure area. If so, then how will this shim layer which is in non-secure area call that function directly instead of through a service?

Reply:
Hi Michael,
You're right. The "mbedtls_mpi_read_binary()" is for key format conversion. It's indeed in the non-secure side as the helper function.
For our shim layer, it's basically:
* Maps PKCS#11 object storage (just certificate and key) to PSA Storage API (the backend of PSA Storage API is TF-M storage service in the example)
* Maps PKCS#11 Crypto API (just the ones used by AWS TLS library, mainly sign function) to PSA Crypto API (the backend of PSA Crypto API is TF-M crypto service which uses its own mbedcrypto)